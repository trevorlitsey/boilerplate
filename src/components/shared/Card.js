import React from 'react';
import { string, object, bool } from 'prop-types';
import { Link } from 'react-router-dom';

const grid = 8;

const getStyles = (overrideStyles, isActive) => ({
	userSelect: 'none',
	padding: grid * 2,
	lineHeight: 1,
	margin: `0 0 ${grid}px ${grid}px`,
	background: isActive ? 'var(--primary)' : '',
	color: isActive ? 'var(--white)' : '',

	...overrideStyles,
})

const Tag = ({ value }) => (
	<span className="badge badge-light">{value}</span>
)

const Card = ({ title, text, tags, id, overrideStyles, isActive, showModal }) => (
	<div className="card" style={getStyles(overrideStyles, isActive)}>
		<h5 className="card-title">{title}</h5>
		<p className="card-text">{text.length > 80 ? text.substring(0, 80) + ' ...' : text}</p>
		<div style={{ marginTop: 'auto', display: 'flex', justifyContent: 'space-between' }}>
			<Link to="#" style={{ marginRight: 12, marginTop: 'auto' }} onClick={() => showModal(id)}>edit</Link>
			<div>
				{tags && tags.map(tag => <Tag key={tag.id} value={tag.value} />)}
			</div>
		</div>
	</div>
)

Card.propTypes = {
	title: string.isRequired,
	text: string.isRequired,
	overrideStyles: object,
	isActive: bool,
}

export default Card;