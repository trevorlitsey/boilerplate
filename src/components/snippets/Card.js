import React from 'react';
import { string, object, bool, array, func } from 'prop-types';
import { Link } from 'react-router-dom';

const grid = 8;

const styles = {
	userSelect: 'none',
	padding: grid * 2,
	lineHeight: 1,
}

export const Tag = ({ value }) => (
	<span className="badge badge-light">{value}</span>
)

const Card = ({ title, text, tags, id, showModal }) => (
	<div className="card" style={styles}>
		<h5 className="card-title">{title}</h5>
		<p className="card-text">{text.length > 80 ? text.substring(0, 80) + ' ...' : text}</p>
		<div style={{ marginTop: 'auto', display: 'flex', justifyContent: 'space-between' }}>
			<a href="#" style={{ marginRight: 12, marginTop: 'auto' }} onClick={() => showModal(null, id)}>edit</a>
			<div>
				{tags && tags.map(tag => <Tag key={tag.id} value={tag.value} />)}
			</div>
		</div>
	</div>
)

Card.propTypes = {
	title: string.isRequired,
	text: string.isRequired,
	tags: array,
	id: string.isRequired,
	showModal: func.isRequired,
}

export default Card;