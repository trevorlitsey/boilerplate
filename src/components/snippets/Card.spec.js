import React from 'react';
import { shallow } from 'enzyme';

import Card, { Tag } from './Card';

describe('Card', () => {

	it('should render without crashing', () => {
		const wrapper = renderCard({ isActive: false });
		expect(wrapper).toMatchSnapshot();
	})

	it('should display title', () => {
		const title = 'this is the title';
		const wrapper = renderCard({ title });
		expect(wrapper.find('.card-title').text()).toBe(title);
	})

	it('should ellipsize text if > 80 characters', () => {
		const longText = Array.from({ length: 81 }, item => 'a').join('');
		const wrapper = renderCard({ text: longText });
		expect(wrapper.find('.card-text').text()).toContain('...');
	})

	it('should display all text if <= 80 characters', () => {
		const almostLongText = Array.from({ length: 80 }, item => 'a').join('');
		const wrapper = renderCard({ text: almostLongText });
		expect(wrapper.find('.card-text').text()).toEqual(almostLongText);
		expect(wrapper.find('.card-text').text()).not.toContain('...');
	})

	it('should display correct number of tags given', () => {
		const tags = [
			{ id: 'jgtv2988', label: 'so', value: 'so' },
			{ id: 'jgtv29wa', label: 'many', value: 'many' },
			{ id: 'jgtv1e21', label: 'tagz', value: 'tagz' },
		]
		const wrapper = renderCard({ tags });
		expect(wrapper.find(Tag).length).toEqual(tags.length)
	})

	it('should call showModal on edit click', () => {
		const showModal = jest.fn();
		const id = 'id'
		const wrapper = renderCard({ showModal, id });
		wrapper.find('a').simulate('click');
		expect(showModal).toHaveBeenCalledWith(null, id);
	})

})

describe('Tag', () => {

	it('should display value given', () => {
		const value = 'what a value!';
		const wrapper = renderTag({ value });
		expect(wrapper.text()).toEqual(value);
	})

})

// -------
function renderCard(props = {}) {
	const propsToUser = {
		title: 'TITLE!',
		text: 'TEXT!',
		tags: [
			{ id: 'jgtv2988', label: 'lots', value: 'lots' },
			{ id: 'jgtv29wa', label: 'of', value: 'of' },
			{ id: 'jgtv1e21', label: 'tags', value: 'tags' },
		],
		id: '567y2hs7',
		overrideStyles: {},
		isActive: false,
		showModal: () => { },
		...props,
	}

	return shallow(<Card {...propsToUser} />)
}

function renderTag(props = {}) {
	const propsToUser = {
		value: 'this is the value',
		...props,
	}

	return shallow(<Tag {...propsToUser} />)
}