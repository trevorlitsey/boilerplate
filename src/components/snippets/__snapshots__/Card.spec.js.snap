// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    id="567y2hs7"
    isActive={false}
    overrideStyles={Object {}}
    showModal={[Function]}
    tags={
      Array [
        Object {
          "id": "jgtv2988",
          "label": "lots",
          "value": "lots",
        },
        Object {
          "id": "jgtv29wa",
          "label": "of",
          "value": "of",
        },
        Object {
          "id": "jgtv1e21",
          "label": "tags",
          "value": "tags",
        },
      ]
    }
    text="TEXT!"
    title="TITLE!"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h5
          className="card-title"
        >
          TITLE!
        </h5>,
        <p
          className="card-text"
        >
          TEXT!
        </p>,
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
              "marginTop": "auto",
            }
          }
        >
          <a
            href="#"
            onClick={[Function]}
            style={
              Object {
                "marginRight": 12,
                "marginTop": "auto",
              }
            }
          >
            edit
          </a>
          <div>
            <Tag
              value="lots"
            />
            <Tag
              value="of"
            />
            <Tag
              value="tags"
            />
          </div>
        </div>,
      ],
      "className": "card",
      "style": Object {
        "lineHeight": 1,
        "padding": 16,
        "userSelect": "none",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "TITLE!",
          "className": "card-title",
        },
        "ref": null,
        "rendered": "TITLE!",
        "type": "h5",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "TEXT!",
          "className": "card-text",
        },
        "ref": null,
        "rendered": "TEXT!",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <a
              href="#"
              onClick={[Function]}
              style={
                Object {
                  "marginRight": 12,
                  "marginTop": "auto",
                }
              }
            >
              edit
            </a>,
            <div>
              <Tag
                value="lots"
              />
              <Tag
                value="of"
              />
              <Tag
                value="tags"
              />
            </div>,
          ],
          "style": Object {
            "display": "flex",
            "justifyContent": "space-between",
            "marginTop": "auto",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "edit",
              "href": "#",
              "onClick": [Function],
              "style": Object {
                "marginRight": 12,
                "marginTop": "auto",
              },
            },
            "ref": null,
            "rendered": "edit",
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Tag
                  value="lots"
                />,
                <Tag
                  value="of"
                />,
                <Tag
                  value="tags"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "jgtv2988",
                "nodeType": "function",
                "props": Object {
                  "value": "lots",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "jgtv29wa",
                "nodeType": "function",
                "props": Object {
                  "value": "of",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "jgtv1e21",
                "nodeType": "function",
                "props": Object {
                  "value": "tags",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h5
            className="card-title"
          >
            TITLE!
          </h5>,
          <p
            className="card-text"
          >
            TEXT!
          </p>,
          <div
            style={
              Object {
                "display": "flex",
                "justifyContent": "space-between",
                "marginTop": "auto",
              }
            }
          >
            <a
              href="#"
              onClick={[Function]}
              style={
                Object {
                  "marginRight": 12,
                  "marginTop": "auto",
                }
              }
            >
              edit
            </a>
            <div>
              <Tag
                value="lots"
              />
              <Tag
                value="of"
              />
              <Tag
                value="tags"
              />
            </div>
          </div>,
        ],
        "className": "card",
        "style": Object {
          "lineHeight": 1,
          "padding": 16,
          "userSelect": "none",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "TITLE!",
            "className": "card-title",
          },
          "ref": null,
          "rendered": "TITLE!",
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "TEXT!",
            "className": "card-text",
          },
          "ref": null,
          "rendered": "TEXT!",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <a
                href="#"
                onClick={[Function]}
                style={
                  Object {
                    "marginRight": 12,
                    "marginTop": "auto",
                  }
                }
              >
                edit
              </a>,
              <div>
                <Tag
                  value="lots"
                />
                <Tag
                  value="of"
                />
                <Tag
                  value="tags"
                />
              </div>,
            ],
            "style": Object {
              "display": "flex",
              "justifyContent": "space-between",
              "marginTop": "auto",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "edit",
                "href": "#",
                "onClick": [Function],
                "style": Object {
                  "marginRight": 12,
                  "marginTop": "auto",
                },
              },
              "ref": null,
              "rendered": "edit",
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Tag
                    value="lots"
                  />,
                  <Tag
                    value="of"
                  />,
                  <Tag
                    value="tags"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "jgtv2988",
                  "nodeType": "function",
                  "props": Object {
                    "value": "lots",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "jgtv29wa",
                  "nodeType": "function",
                  "props": Object {
                    "value": "of",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "jgtv1e21",
                  "nodeType": "function",
                  "props": Object {
                    "value": "tags",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
